services_app/src/main/java/org/opendatakit/services/preferences/fragments/ServerSettingsFragment.java

```diff
@@ -273,24 +305,171 @@ public boolean onPreferenceChange(Preference preference, Object newValue) {
+          toastString +="URL : "+url+"\n";
+        }
+      }
+      catch (Exception e) {
+        toastString +="URL : not found\n";
+        Log.i(TAG,"Url not found");
+      }
+      try{
+        String username = (String) mainObject.get("username");
+        if(username!=null && (usernamePasswordAvailable || adminMode)){
+          onPreferenceChange(mUsernamePreference,username);
+         if(credentialAvailable || adminMode) signOnPreferenceChanged(mSignOnCredentialPreference,"username_password");
+          toastString = toastString +"Username : "+username+"\n";
+        }
+      }
+      catch (Exception e) {
+        toastString +="Sign-on Credential type: anonymous\n";
+        toastString +="Username : not found\n";
+        if(credentialAvailable || adminMode) signOnPreferenceChanged(mSignOnCredentialPreference,"none");
+        Log.i(TAG,"Username not found");
+      }
+
+     try{
+       String password = (String) mainObject.get("password");
+       if(password!=null && (usernamePasswordAvailable || adminMode)){
+         updatePassword(password);
+         toastString +="Password : "+password;
+       }
+     }
+     catch (Exception e) {
+       toastString +="Password : not found";
+       Log.i(TAG,"Password not found");
+     }
+
+     Toast.makeText(getActivity(), toastString, Toast.LENGTH_SHORT).show();
+
+    } catch (JSONException e) {
+      Toast.makeText(getActivity(), "Invalid Qr code.", Toast.LENGTH_SHORT).show();
+      Log.i(TAG,"Invalid QR code");
+    }
+
+  }
+
+  public void updatePassword(String pw) {
+    updatePreferenceSummary(CommonToolProperties.KEY_PASSWORD, pw);
+ }
+
+  @Override
+  public boolean onOptionsItemSelected(MenuItem item) {
+    switch (item.getItemId()) {
+      case R.id.action_barcode:
+        // When Scan QR icon is clicked.
+        if (ActivityCompat.checkSelfPermission(getActivity(), Manifest.permission.CAMERA)
+                == PackageManager.PERMISSION_GRANTED) {
+          // Permission is already available, start camera preview
+           openBarcodeScanner();
+        } else {
+          // Permission is missing and must be requested.
+          requestCameraPermission();
+        }
+        return true;
+
+      default:
+        return super.onOptionsItemSelected(item);
+
+    }
+  }
+
+  private void openBarcodeScanner() {
+    IntentIntegrator.forFragment(this)
+        .setDesiredBarcodeFormats(IntentIntegrator.QR_CODE_TYPES)
+        .setPrompt("Place QR Code inside the rectangle")
+        .setCameraId(0)
+        .setBeepEnabled(true)
+        .setBarcodeImageEnabled(false)
+        .initiateScan();
+  }
+
+  @Override
+  public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions,
+                                         @NonNull int[] grantResults) {
+      super.onRequestPermissionsResult(requestCode,permissions,grantResults);
+     if (requestCode == PERMISSION_REQUEST_CAMERA) {
```

Comment:
We already have a utility method that handles permissions, please use it. 
See BaseLauncherActivity in androidcommon for reference, or TableDisplayActivity in Tables.

Reply:
You're right, I didn't notice that. We request permission from the activity everywhere else. Let's just leave it as is, it's not that big of a problem.